### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by apexdodge** [RaythaHQ/raytha#172] [Ideas]
Brainstorming the roadmap. Below are a bunch of great ideas with varying levels of engineering difficulty and impact. No particular order.

**Future feature ideas**

difficulty level: easy
- Parent-child site pages relationship
- Allow site pages to be of type [page, page group (no content), 301 redirect]
- Shared functions (functions that can be called by other functions)
- Archive content types

difficulty level: medium
- `render_widget()` liquid template helper to render one-offs in a template
- Custom widgets
- Media management
- Additional constraints on custom fields, like min-max values
- MCP

difficulty level: hard
- Import content types
- Intake Forms
- Full Text Search
- Multi-lingual
- Multi-site
- Multi-tenant

**Completed**
- ✅ **Update default theme**. Current default theme is not great. We should just go with the most vanilla, out of the box bootstrap theme we can offer, so it is a blank canvas for anyone who wants to get started quickly. (released v1.1.5) #174 
- ✅ **Embedded plugins / Scripting language**. The initial implementation will support triggers from HTTP get and post requests. (released v1.2.0) #177 
- ✅ **Navigation helper** - An admin UI to allow website maintainers to create different menus that can then be used in the themes to dynamically update menus. (released v1.3.0) #182 
- ✅ **Web hooks / triggers** - Coming in version 1.3.0. We are expanding the `Raytha Functions` capability to support additional triggers, such as when a content item is updated. This enhancement allows the user to create webhooks (released v1.3.0) #82 
- ✅ **Multiple Themes**. Have different themes installed, Wordpress style, and then you can choose to activate a particular theme, or switch to a different theme. Right now, we only support 1 set of templates. This would also allow people to install a theme from a URL, which opens up possibilities for a theme gallery or marketplace. (released v1.3.0)
- ✅ **TinyMCE replacement** - TinyMCE is no longer MIT licensed, so we will roll our UI on top of the headless wysiwyg editor TipTap #196 (released v1.4.1)
- ✅ **Monaco replacement** - Monaco editor is being replaced with CodeMirror6 to speed up build times and cut the bundled javascript size in half. #228 (released v1.4.1)
- ✅ **Postgres** - Postgres became a supported database in addition to MS SQL Server. #226 (released v1.4.1)
- ✅ **Swagger UI replacement** - SwaggerUI has been replaced with Scalar (released v1.4.2)
- ✅ **Maintenance feature** - Ability to clear audit logs (released v1.4.3)
- ✅ **Health endpoint** - Can now go to /healthz for a healthcheck (released v1.4.3)
- ✅ **AI Assisted Theme Development** - Separate project, compatible with v1.4.6+.[ Work with Cursor to speed up raytha theme development](https://github.com/raythahq/theming-with-cursor/) (released v1.4.6)
- ✅ **Brute force protection on login** - Configurable options for timeout window for failed login attempts (released v1.5.0)
- ✅ **Drag & Drop Page Builder w/ widgets** - (released v.1.5.0)

**Comment by ridingmower** [comment]
Some ideas/features/Suggestions (Which maybe questions and food for thought)

- [ ] In place upgrades to the CMS, similar to WP
- [ ] separating developer code, is this a widget or a modules or a new plugin
- [ ] Template generator for the developer for the backend files needed, and views as well.
- [ ] Media support, manage your pictures, songs, etc. And allow us to have a simple grid gallery
- [ ] Observability to know any issues/hickups in production, memory leaks
- [ ] Data protection, Roles are great, but when you have two managers from different companies how to firewall that data, is it a claim/group?

**Comment by apexdodge** [comment]
v1.3.0 will be a huge release and will be publicly available the next week.

Includes:

- Navigation helper
- Web hooks / triggers
- Support for multiple themes
